ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java

```diff
@@ -1963,33 +1962,24 @@ private static Set<Operator<?>> findParentWorkOperators(ParseContext pctx,
     return set;
   }
 
-  private static Set<Operator<?>> findAscendantWorkOperators(ParseContext pctx,
-          SharedWorkOptimizerCache optimizerCache, Operator<?> start) {
-    // Find operators in work
-    Set<Operator<?>> workOps = findWorkOperators(optimizerCache, start);
-    // Gather input works operators
-    Set<Operator<?>> result = new HashSet<Operator<?>>();
-    Set<Operator<?>> set;
-    while (!workOps.isEmpty()) {
-      set = new HashSet<Operator<?>>();
-      for (Operator<?> op : workOps) {
-        if (op.getParentOperators() != null) {
-          for (Operator<?> parent : op.getParentOperators()) {
-            if (parent instanceof ReduceSinkOperator) {
-              set.addAll(findWorkOperators(optimizerCache, parent));
-            }
-          }
-        } else if (op instanceof TableScanOperator) {
-          // Check for DPP and semijoin DPP
-          for (Operator<?> parent : optimizerCache.tableScanToDPPSource.get((TableScanOperator) op)) {
-            set.addAll(findWorkOperators(optimizerCache, parent));
-          }
+  private static Set<Operator<?>> findAscendantWorkOperators(SharedWorkOptimizerCache optimizerCache,
```

Comment:
do I understand it right that original `findWorkOperators` collects recursively both operator's parents and children starting from `start`, but we only interested in parents?

Note findWorkOperators leverages the cache, do we need the same here?

Reply:
1 .
Yes, `findWorkOperators` returns the set of operators that will be packed in to the same work (`Vertex` in Tez and `BaseWork` in Hive).
Using `findWorkOperators`, we can compute the set of operators that belongs to the same work or its ancestor work. But this set is not our interest because the dependency between works does not always imply the dependency between operators.

For example, we can find 5 works from the below operator graph. (The same as the one in the slide)
```
Work1: MAPJOIN_217, SEL_27, GBY_29, SEL_140, GBY_141, RS_142, SEL_185, GBY_186, RS_187
Work2: GBY_143, RS_144
Work3: TS_3, FIL_116
Work4: GBY_188, RS_189
Work5: TS_41, FIL_121
```
In the example, RS_189 belongs to Work4, which depends on Work1. But RS_189 does not depend on RS_142, which is a member of Work1.

2 .
Since **the set of ascendant operators** is used less frequently than **the set of operators that belong to the same work**, I think we don't need another cache that is dedicated to `findAscendantWorkOperators`.

3 .
We don't use `findParentWorkOperators` anymore, and I don't know its original usage. Since findParentWorkOperators is just a one-step version of findAscendantWorkOperators, there might be the same issue if one used this method to find any dependency conflicts between operators.

4 .
I have changed the name of the method to `findAscendantOperators` as the method does not use the concept of work here.

![h](https://github.com/user-attachments/assets/682c5659-e708-4650-bbbf-187ea7d5d381)

Reply:
👍 , thanks for the explanation @ngsg!